### ISSUE TITLE
Consider using `ragged` arrays to handle `cls` data

### ISSUE THREAD

**Issue by paddyroddy** [glass-dev/glass#382]
### Is Your Feature Request Related to a Problem? Please Describe

https://github.com/scikit-hep/ragged is designed to handle ragged arrays in an Array API compliant manner, i.e.

```python
>>> import ragged
>>> a = ragged.array([[[1.1, 2.2, 3.3], []], [[4.4]], [], [[5.5, 6.6, 7.7, 8.8], [9.9]]])
>>> a
ragged.array([
    [[1.1, 2.2, 3.3], []],
    [[4.4]],
    [],
    [[5.5, 6.6, 7.7, 8.8], [9.9]]
])
```

**Comment by paddyroddy** [comment]
Moving to blocked due to uncertainty behind #403, more specifically the use of `s2ftt`.

**Comment by paddyroddy** [comment]
@Saransh-cpp you have more experience with `ragged`. Do you have any chance to make a comment about this? From the investigation in #973, it looks like it might be possible to use `ragged`.

**Comment by Saransh-cpp** [comment]
I think I have a solution.

Given that Awkward takes the underlying NumPy array as a view, it can be mutated (through the NumPy array, as in my comment above), but, given that Awkward copies Python objects (such as a List of NumPy arrays), they are pretty much immutable. However, you can go into Ragged's (Awkward's) low-level API to actually mutate the array:

```py
a = [np.array([1, 2, 3]), np.array([1, 2])]
r = ragged.asarray(a)

# r -> ragged.array([
#     [1, 2, 3],
#     [1, 2]
# ])

# will give you the upstream Awkward array - Ragged arrays are just specialised Awkward arrays
r._impl
# <Array [[1, 2, 3], [1, 2]] type='2 * var * int64'>

# the XML-like representation of an Awkward array
# notice how the content is contiguous, and the offsets store the how to look at it in the memory (giving that ragged functionality to the user)
r._impl.layout
# <ListOffsetArray len='2'>
#     <offsets><Index dtype='int64' len='3'>[0 3 5]</Index></offsets>
#     <content><NumpyArray dtype='int64' len='5'>[1 2 3 1 2]</NumpyArray></content>
# </ListOffsetArray>

# and finally mutate the array
# you will need to use `r._impl.layout.offsets.data` to navigate to specific indices of the array if you want to maintain the nice indexing syntax
r._impl.layout.content.data[3] = 999
# r -> ragged.array([
#     [1, 2, 3],
#     [999, 2]
# ])
```

I would recommend keeping Cls as a List of Lists if you want to only use Ragged for Cls computations as Ragged can only take NumPy arrays and Lists as inputs. It can take JAX arrays as well, but the backend is deprecated and will be removed soon. In any case, given that the computations are performed by Ragged, they will be fast and not slowed down by the input being a Python List of Lists. Also, given that Ragged is an array producer, it is very well not required to be compatible with JAX or Dask (that is, accept JAX or Dask arrays as inputs), but it is only required to keep its API uniform to them.

Using only List of Lists:

```py
In [16]: a = [[1, 2, 3], [1, 2]]

In [17]: r = ragged.asarray(a)

In [18]: r._impl.layout.offsets.data
Out[18]: array([0, 3, 5])

In [19]: r._impl.layout.content.data[3] = 999

In [20]: r
Out[20]: 
ragged.array([
    [1, 2, 3],
    [999, 2]
])
```

I think it should be easy to create a wrapper `at` method that dispatches `xpx`'s `at` if the backend is not Ragged and use the code above if the backend is Ragged. I hope this helps! 🙂 

If adopted - would speed up Cls computations as Ragged matches the speed of NumPy (C++ underneath) + make the computations easy as Ragged follows the NumPy syntax (no more dealing with a List of arrays) and array API; however, will come at a cost of maintaining a new `at` wrapper (and possibly more wrappers down the line).

**Comment by Saransh-cpp** [comment]
Using `at` for mutation is now possible with ragged arrays - https://github.com/scikit-hep/ragged/pull/128 👀

**Comment by paddyroddy** [comment]
@Saransh-cpp I gave it another go and wasn't getting far https://github.com/glass-dev/glass/pull/1109. Any feedback, suggestions or contributions welcome!

**Comment by paddyroddy** [comment]
```python
import jax.numpy as jnp
import numpy as np
import ragged

jax.config.update('jax_enable_x64', True)

ragged.asarray([np.asarray([1,2,3])]) # works
ragged.asarray([jnp.asarray([1,2,3])]) # JAX features require `ak.jax.register_and_check()`
```

**Comment by ntessore** [comment]
Yeah I think this is not going to work with `ragged` unless it itself supported arbitrary array backends. So while the goal is still to move to a more intuitive representation of Cls (and alms) I'm not sure this here is the way to go.

**Comment by Saransh-cpp** [comment]
Ah yes, as I mentioned above:

> I would recommend keeping Cls as a List of Lists if you want to only use Ragged for Cls computations as Ragged can only take NumPy arrays and Lists as inputs. It can take JAX arrays as well, but the backend is deprecated and will be removed soon.
> ...
> Also, given that Ragged is an array producer, it is very well not required to be compatible with JAX or Dask (that is, accept JAX or Dask arrays as inputs), but it is only required to keep its API uniform to them.

If the end goal is to have Ragged consume JAX arrays, then this is not feasible (as Ragged is more of a provider library, and less of a consumer library) and we can close this issue

**Comment by paddyroddy** [comment]
Thanks for clarifying @Saransh-cpp, will close.
